consensus.go

```diff
@@ -0,0 +1,181 @@
+	}
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	// The wallet will need to be unlocked to sign the API
+	// request(s) for setting this vote choice with the VSP.
+	err := wallet.UnlockWallet(passphrase)
+	if err != nil {
+		return translateError(err)
+	}
+	defer wallet.LockWallet()
+
+	ctx := wallet.shutdownContext()
+
+	// get choices
+	choices, _, err := wallet.Internal().AgendaChoices(ctx, ticketHash) // returns saved prefs for current agendas
+	if err != nil {
+		return err
+	}
+
+	currentChoice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: "abstain", // default to abstain as current choice if not found in wallet
+	}
+
+	for i := range choices {
+		if choices[i].AgendaID == agendaID {
+			currentChoice.ChoiceID = choices[i].ChoiceID
+			break
+		}
+	}
+
+	newChoice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, newChoice)
+	if err != nil {
+		return err
+	}
+
+	// If ticket hash is provided, then set vote choice for the selected ticket
+	// with the associated vsp.
+	if ticketHash != nil {
+		vspTicketInfo, err := wal.VSPTicketInfo(ctx, ticketHash)
+		if err != nil {
+			return err
+		}
+		// Update the vote choice for the ticket with the vsp
+		vspClient, err := wallet.VSPClient(vspTicketInfo.Host, vspTicketInfo.PubKey)
+		if err != nil {
+			return err
+		}
+		err = vspClient.SetVoteChoice(ctx, ticketHash, newChoice)
+		if err != nil {
+			// Updating the agenda voting preference with the vsp failed, therefore
+			// revert the locally saved voting preference for the agenda.
+			_, err = wal.SetAgendaChoices(ctx, ticketHash, currentChoice)
+		}
+		return err
+	}
+
+	vspClient, err := wallet.VSPClient(vspHost, vspPubKey)
+	if err != nil {
+		return err
+	}
+
+	// Ticket hash wasn't provided, therefore set vote choice for all tickets
+	// belonging to the selected wallet and controlled by the specififed VSP
+	var firstErr error
+	vspClient.ForUnspentUnexpiredTickets(ctx, func(hash *chainhash.Hash) error {
+		// Never return errors here, so all tickets are tried.
+		// The first error will be returned to the user.
+		err := vspClient.SetVoteChoice(ctx, hash, newChoice)
+		if err != nil && firstErr == nil {
+
+			firstErr = err
+		}
+		return nil
+	})
+
+	if firstErr != nil {
+		// Updating the agenda voting preference with the vsp failed, therefore
+		// revert the locally saved voting preference for the agenda.
+		_, err = wal.SetAgendaChoices(ctx, ticketHash, currentChoice)
+		log.Error(err)
+	}
+
+	return firstErr
+}
+
+// AllVoteAgendas returns all agendas through out the various
+// stake versions for the active network and this version of the software,
+// as well as returns the vote preferences for the current version,
+// and all agendas defined by it for a selected wallet.
```

Comment:
This needs cleanup. Particularly the irregular characters-per-line length but also the last sentence looks off for me - don't understand what is meant there or how that sentence plays out with the rest of the comment.

I recommended:
```
// AllVoteAgendas returns all agendas of all stake versions for the active
// network and this version of the software. Also returns any saved vote
// preferences for the agendas of the current stake version. Vote preferences
// for older agendas cannot currently be retrieved.
```